### ISSUE TITLE
False positive 'unpacking-non-sequence' - probably the issue is in analyzing 'structural matching' pattern.

### ISSUE THREAD

**Issue by Questionsdenombres** [PyCQA/pylint#7316]
### Bug description

```python
from enum import Enum, auto

class X(Enum):
    X1 = auto()
    X2 = auto()
    X3 = auto()

def X_func(x: X) -> tuple[int, str]:
    match x:
        case X.X1:
            return (1, 'X1')
        case X.X2:
            return (2, 'X2')
        case X.X3:
            return (3, 'X3') 
            # end #1
        case _:
            raise ValueError
            # end #2
    raise ValueError
    # end #3
    # return (0, '')
    # end #4

(a, b) = X_func(X.X1) # here is the pylint reported error.
```

### Command used

pylint launched by "Python extension for Visual Studio Code" (v2022.12.0)

### Pylint output

pylint reports an error on the last line:
- unpacking-non-sequence (Attempting to unpack a non-sequence)
- Note that pylint doesn't see that all lines of `X_func(...)` after `end #1` are actually unreachable.
- If you uncomment the line `return (0, '')` the error disappears and 
  pylint now sees that this uncommented line is unreachable.

### Expected behavior

No reported error.

My analysis:
- pylint should see that all lines after `# end #2` are unreachable.
- But, I am unsure whether pylint should see that all lines after `# end #1` are unreachable.
  Because this looks a type checking feature: a type checker should see it; pylance/pyright does.

Thank you for your great job!
Patrick Bourdon - PARIS.

### Pylint version
- "Python extension for Visual Studio Code" (v2022.12.0 that uses pylint 2.14.5)
- Python version(3.10.6 (tags/v3.10.6:9c7b4bd, Aug  1 2022, 21:53:49) [MSC v.1932 64 bit (AMD64)])
- astroid 2.11.7

### OS / Environment

- Windows-10-10.0.22000-SP0
- Visual studio: 1.70.1

**Comment by jacobtylerwalls** [comment]
Verified this is resolved by PyCQA/astroid#2042, so closing as a duplicate of #5288.
